storage-proofs-porep/src/stacked/vanilla/proof.rs

```diff
@@ -928,43 +963,69 @@ impl<'a, Tree: 'static + MerkleTreeTrait, G: 'static + Hasher> StackedDrg<'a, Tr
+                let mut tree_builder = None;
+                let mut is_final = false;
+                let mut i = 0;
+                let call = |_: usize, alloc: Option<&ResourceAlloc>| -> Result<TaskResult, Error> {
+                    if tree_builder.is_none() {
+                        let alloc = alloc.ok_or(Error::Unclassified(
+                            "No resource allocation for TreeBuilder".to_string(),
+                        ))?;
+                        let device_hash = alloc.resource_id[0];
+                        let device = get_device_by_hash(device_hash).ok_or(Error::Unclassified(
+                            "No resource allocation for TreeBuilder".to_string(),
+                        ))?;
+                        let selector = GPUSelector::UniqueId(device.device_id().ok_or(
+                            Error::Unclassified(
+                                "No resource allocation for TreeBuilder".to_string(),
+                            ),
+                        )?);
+                        tree_builder = Some(
+                            TreeBuilder::<Tree::Arity>::new(
+                                Some(BatcherType::CustomGPU(selector)),
+                                nodes_count,
+                                max_gpu_tree_batch_size,
+                                tree_r_last_config.rows_to_discard,
+                            )
+                            .map_err(|e| Error::Unclassified(e.to_string()))?,
                         );
-                        let (_, tree_data) = tree_builder
-                            .add_final_leaves(&encoded)
-                            .expect("failed to add final leaves");
+                    }
+                    let builder = tree_builder.as_mut().ok_or(Error::Unclassified(
+                        "No resource allocation for tree builder".to_string(),
+                    ))?;
 
-                        writer_tx.send(tree_data).expect("failed to send tree_data");
-                        break;
+                    // Loop until all trees for all configs have been built.
+                    if i < config_count {
+                        if !is_final {
+                            let (encoded, last) = builder_rx
+                                .recv()
+                                .map_err(|e| Error::Unclassified(e.to_string()))?;
+
+                            // Just add non-final leaf batches.
+                            builder.add_leaves(&encoded)?;
+                            is_final = last;
+                        } else {
+                            // If we get here, this is a final leaf batch: build a sub-tree.
+                            info!(
+                                "building base tree_r_last with GPU {}/{}",
+                                i + 1,
+                                tree_count
+                            );
+                            if let Some((_, tree_data)) = builder.build_next()? {
```

Comment:
The comment from above applies here too, except instead of a missing call to `add_columns()`, there is a missing call to `add_leaves()`. The same fix and same recommendation for a corresponding enhancement to the new `neptune` API apply. Making the `neptune` change will mitigate the increased chance for mistake (as seen here) introduced by removing the need for the caller to explicitly add the final leaves/columns.

Reply:
The above correction applies here too: https://github.com/filecoin-project/rust-fil-proofs/pull/1460/files#r631920821